### PR TITLE
Find on shelf modal

### PR DESCRIPTION
**Pull request by Adamik10**
#### Ticket
https://reload.atlassian.net/browse/DDFSOEG-227

#### Description
This PR adds the find on shelf modal and as a part of it:
- new disclosure variant - with availability label
- new availability label variant - without manifestation type
- Find on shelf list component, its story and styling
- Find on shelf modal component, its story and styling

#### Screenshots
Disclosure with availability label:
image
image

Availability label without manifestation type:
image

Find on shelf list:
image
image

Find on shelf modal:
image
image

#### Additional comments
I disabled the controls for Availability label because all the different controls are covered in the specific stories and it can get  pretty confusing switching between the stories and seeing cached availability labels that don't correspond to the story.

### CODE REVIEW

File: src/stories/Library/Lists/list-find-on-shelf/ListFindOnShelf.tsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      <li className="find-on-shelf__header-row text-small-caption">
+        <span>Materiale</span>
+        <span>Find det på hylden</span>
+        <span>Hjemme</span>
+      </li>
+      {numberArray.map((key) => {
+        return (
+          <li className="find-on-shelf__row text-body-medium-regular" key={key}>
+            <span>{manifestationName}</span>
+            <span>{location}</span>
```
Comment (kasperg) [inline src/stories/Library/Lists/list-find-on-shelf/ListFindOnShelf.tsx:27]: Consider whether the location actually consists of multiple DOM elements with decorative separators (implemented in CSS) instead of being a single text string.

File: src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+          margin-right: $s-xl;
+          overflow-y: visible;
+        }
+      }
+
+      & .hide-on-desktop {
+        @include breakpoint-s {
+          display: none;
+        }
+      }
```
Comment (kasperg) [inline src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss:84]: Please restructure this.

It is not very BEM like to combine `.find-on-shelf` with another class outside of the block. I suggest making `.hide-on-desktop` a utility class like `.hide-visually` in `shared.css`. An option might be to add a modifer to `.hide-visually` which only applies on desktop.

File: src/stories/Library/Modals/modal-find-on-shelf/ModalFindOnShelf.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        return (
+          <Disclosure
+            withAvailability
+            headline="Bibliotek fliale navn"
+            icon="Various"
+            key={branchKey}
+          >
+            <ul className="find-on-shelf">
+              <li className="find-on-shelf__header-row text-small-caption">
+                <span>Materiale</span>
```
Comment (kasperg) [inline src/stories/Library/Modals/modal-find-on-shelf/ModalFindOnShelf.tsx:65]: Please reuse the `ListFindOnShelf` component programmatically if that is what you intend.

File: src/stories/Library/Modals/modal-find-on-shelf/modal-find-on-shelf.scss [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    @include breakpoint-m {
+      margin: 96px 0 $s-sm 16px;
+    }
+  }
+
+  .disclosure {
+    width: 100%;
+    margin: 0;
+  }
```
Comment (kasperg) [inline src/stories/Library/Modals/modal-find-on-shelf/modal-find-on-shelf.scss:29]: Please reconsider the BEM structure.

If we want different styling for the `.disclosure` then it should be a modifier for that block instead of nesting different blocks.

File: src/stories/Library/availability-label/AvailabilityLabel.tsx [hunk 5]
```diff
@@ -2,8 +2,11 @@ import { AvailabilityLabelPropsType } from "../../availability-label/types";
 import { Pagefold } from "../pagefold/Pagefold";
 import { withAvailabilityProps } from "./abilityLabel.hoc";
 
-export const AvailabilityLabel = (props: AvailabilityLabelPropsType) => {
-  const { manifestation, availability, status } = props;
+export const AvailabilityLabel: React.FC<AvailabilityLabelPropsType> = ({
```
Comment (kasperg) [inline src/stories/Library/availability-label/AvailabilityLabel.tsx:5]: Consider if something is missing here. It does not seem as if we import `React.FC`. 

I wonder why this change is needed anyway. The version before was shorter 🤔 .

File: src/stories/Library/disclosure/Disclosure.tsx [hunk 6]
```diff
@@ -20,16 +23,26 @@ export const Disclosure: React.FC<DisclosureProps> = ({
...
+              src={`icons/collection/${icon}.svg`}
+              alt="various-icon"
+            />
+          </div>
+        )}
+        {!withAvailability && (
+          <span className="disclosure__headline__text">{headline}</span>
+        )}
+        {withAvailability && (
+          <span className="disclosure__headline__text-shorter">{headline}</span>
```
Comment (kasperg) [inline src/stories/Library/disclosure/Disclosure.tsx:39]: Please rework this BEM structure. You should not have nested elements (headline and text) within your block (disclosure). Modifiers should be separated with `--`.

**Review by kasperg — CHANGES_REQUESTED** [review]
:+1: Lots of code, lots of comments.

There are a number of comments regarding the BEM structure of the implementation. I am also not sure how I feel about the `> span` and `:first/last-of-type` rules here compared to using actual BEM elements and modifiers. Some stuff may be easier to go through face to face or in an online meeting.

File: src/stories/Library/Lists/list-find-on-shelf/ListFindOnShelf.tsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            <span>
+              {nrAvailable}
+              <span className="hide-on-desktop"> hjemme</span>
+            </span>
+          </li>
+        );
+      })}
+    </ul>
+  );
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/stories/Library/Lists/list-find-on-shelf/ListFindOnShelf.tsx:36]: That is a great suggestion and I have been thinking about it myself. But the mobile is indeed a list, so I think for now I will leave it as is, and take this with me into my future endeavours.

File: src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+        @include breakpoint-s {
+          margin-left: $s-xl;
+        }
+      }
+      &:last-of-type {
+        margin-right: $s-md;
+        margin-left: auto;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss:28]: Both are better implementations but I like the padding idea more :) Thanks for suggestions! This helps a lot!

File: src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+    @include breakpoint-s {
+      flex-direction: row;
+      min-height: 110px;
+      align-items: center;
+      justify-content: flex-start;
+    }
+
+    &:last-of-type {
+      border-bottom: none;
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss:45]: Sounds good! :)

File: src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+    & > span {
+      margin: $s-md $s-md 0 $s-md;
+
+      &:first-of-type {
+        margin-top: $s-lg;
+      }
+
+      &:last-of-type {
+        margin-bottom: $s-lg;
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss]: Yes! :)

File: src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        margin: 0;
+        max-width: 600px;
+        text-overflow: ellipsis;
+        overflow-y: hidden;
+
+        &:first-of-type {
+          margin-left: $s-xl;
+          margin-top: inherit;
+          width: 404px;
+          min-width: 404px;
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss]: These are actually necessary due to flex shrinking and stretching. I commented it in the code and can show the problems with using just one to you in practice if you'd like 😉

File: src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+      align-items: center;
+      justify-content: flex-start;
+    }
+
+    &:last-of-type {
+      border-bottom: none;
+    }
+
+    & > span {
+      margin: $s-md $s-md 0 $s-md;
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss]: Perfect suggestion. Margin is so much more inferior to gap! :)

File: src/stories/Library/Lists/list-find-on-shelf/ListFindOnShelf.tsx (hunk 1 above)
Comment (Adamik10) [reply src/stories/Library/Lists/list-find-on-shelf/ListFindOnShelf.tsx]: The data looks like this: 
image
I can see location, sublocation, and even department that could be a part of this string. However, even in this case we can just use the "centered dot" ( · ) character to separate these when implementing the react components. And it will look exactly like the design brief I believe. 
So great point, but I'll be running onwards with this as is I think.

File: src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss (hunk 2 above)
Comment (Adamik10) [reply src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss]: Great point! It is a utility class now :)

File: src/stories/Library/Modals/modal-find-on-shelf/ModalFindOnShelf.tsx (hunk 3 above)
Comment (Adamik10) [reply src/stories/Library/Modals/modal-find-on-shelf/ModalFindOnShelf.tsx:65]: Oh snap, of course! :) Thanks!

File: src/stories/Library/Modals/modal-find-on-shelf/modal-find-on-shelf.scss (hunk 4 above)
Comment (Adamik10) [reply src/stories/Library/Modals/modal-find-on-shelf/modal-find-on-shelf.scss:29]: That makes total sense. Thanks :)

File: src/stories/Library/availability-label/AvailabilityLabel.tsx (hunk 5 above)
Comment (Adamik10) [reply src/stories/Library/availability-label/AvailabilityLabel.tsx:5]: Importing react is no longer necessary since React 17:
image

And I just changed it for consistency's sake, as that's how we implement our components throughout the project. Note that there might be some leftover older components from before our time where this is not the case, but generally, this way is what I'm used to seeing. :)

File: src/stories/Library/disclosure/Disclosure.tsx (hunk 6 above)
Comment (Adamik10) [reply src/stories/Library/disclosure/Disclosure.tsx:39]: Oh I had no idea that BEM elements don't nest. Great that you point it out :)

File: src/stories/Library/disclosure/disclosure.scss [hunk 13]
```diff
@@ -19,14 +52,26 @@
...
+    margin-right: $s-md;
+
+    @include breakpoint-s {
+      margin-left: -8px;
+      margin-right: $s-lg;
+    }
 
     & > img {
-      width: 24px;
+      width: $s-lg;
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/stories/Library/disclosure/disclosure.scss:63]: Thank you, kind sir :)

File: src/stories/availability-label/types.tsx [hunk 14]
```diff
@@ -1,5 +1,5 @@
 export type AvailabilityLabelPropsType = {
-  manifestation: "Bog" | "Ebog" | "Lydbog (net)" | "Lydbog (cd-mp3)";
+  manifestation?: "Bog" | "Ebog" | "Lydbog (net)" | "Lydbog (cd-mp3)";
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/stories/availability-label/types.tsx]: Uff yes that's absolutely right! I corrected it around the app (hopefully all places 🤞 )

File: src/stories/Library/disclosure/disclosure.scss [hunk 15]
```diff
@@ -1,16 +1,49 @@
...
+
+      @include breakpoint-s {
+        width: 685px;
+      }
+
+      &-shorter {
+        width: 197px;
+        text-overflow: ellipsis;
+        max-height: 84px;
+        overflow-y: hidden;
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/stories/Library/disclosure/disclosure.scss]: Mmmm that leads me to a terrifying conclusion - that I've been using these wrong. 
Better late than never I guess! Thanks for pointing it out 🥇 :)

File: src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    }
+  }
+
+  &__material-text {
+    max-width: 600px;
+
+    @include breakpoint-s {
+      // Both width and min-width need to be here to prevent flex shrink & stretch.
+      width: 404px;
+      min-width: 404px;
```
Comment (kasperg) [inline src/stories/Library/Lists/list-find-on-shelf/list-find-on-shelf.scss:47]: :+1:

File: src/stories/Library/Modals/modal-find-on-shelf/modal-find-on-shelf.scss [hunk 17]
```diff
@@ -0,0 +1,25 @@
+.modal-find-on-shelf {
+  padding: 0;
+  overflow-y: scroll;
+  align-items: flex-start;
+
+  @include breakpoint-m {
+    padding: 0 215px $s-4xl 215px;
```
Comment (kasperg) [inline src/stories/Library/Modals/modal-find-on-shelf/modal-find-on-shelf.scss:7]: We actually have a variable for 215px: `$s-7xl`.

File: src/stories/Library/Modals/modal-find-on-shelf/modal-find-on-shelf.scss [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+    @include breakpoint-m {
+      margin-top: $s-4xl;
+    }
+  }
+
+  &__caption {
+    margin: $s-2xl 0 $s-md $s-md;
+
+    @include breakpoint-m {
+      margin: 96px 0 $s-sm 16px;
```
Comment (kasperg) [inline src/stories/Library/Modals/modal-find-on-shelf/modal-find-on-shelf.scss:22]: We also have a variable for `16px`: `$s-md`.

File: src/stories/Library/disclosure/disclosure.scss [hunk 19]
```diff
@@ -19,14 +58,26 @@
     border-radius: 20px;
     display: flex;
     justify-content: center;
+    margin-right: $s-md;
+
+    @include breakpoint-s {
+      margin-left: -8px;
```
Comment (kasperg) [inline src/stories/Library/disclosure/disclosure.scss:64]: I wonder if we can use a variable here with an expression e.g. `-#{$s-sm}`.

File: src/stories/Library/Modals/modal-find-on-shelf/modal-find-on-shelf.scss (hunk 17 above)
Comment (Adamik10) [reply src/stories/Library/Modals/modal-find-on-shelf/modal-find-on-shelf.scss:7]: This piece of code doesn't exist anymore :) 
But good to know we have, oh, so many variables for sizes 😛

File: src/stories/Library/Modals/modal-find-on-shelf/modal-find-on-shelf.scss (hunk 18 above)
Comment (Adamik10) [reply src/stories/Library/Modals/modal-find-on-shelf/modal-find-on-shelf.scss:22]: This is also outdated 🙈

File: src/stories/Library/disclosure/disclosure.scss (hunk 19 above)
Comment (Adamik10) [reply src/stories/Library/disclosure/disclosure.scss]: Indeed we can do that! Thanks! :)

**Review by sinejespersen — APPROVED** [review]

**Review by kasperg — APPROVED** [review]
